routes/main.js

```diff
@@ -0,0 +1,218 @@
+});
+
+router.get('/products', async (req, res) => {
+	const itemsPerPage = 9;
+	const page = parseInt(req.query.page) || 1;
+	const category = req.query.category;
+	const priceSort = req.query.price;
+	const searchParam = req.query.query;
+
+	let query = {};
+
+	if (category) {
+		query.category = category;
+	}
+
+	if (searchParam) {
+		query.$or = [
+			{ name: { $regex: new RegExp(searchParam, 'i') } }, // Case-insensitive search in names
+			{ category: { $regex: new RegExp(searchParam, 'i') } }, // Case-insensitive search in categories
+		];
+	}
+
+	let sortCriteria = {};
+
+	if (priceSort === 'highest') {
+		sortCriteria = { price: -1 };
+	} else if (priceSort === 'lowest') {
+		sortCriteria = { price: 1 };
+	}
+
+	try {
+		const totalCount = await Product.countDocuments(query);
+		if (totalCount === 0) {
+			return res.status(404).send('No products found');
+		}
+		const products = await Product.find(query)
+			.sort(sortCriteria)
+			.skip((page - 1) * itemsPerPage)
+			.limit(itemsPerPage);
+
+		const totalPages = Math.ceil(totalCount / itemsPerPage);
+
+		res.status(200).json({
+			products,
+			totalPages,
+			currentPage: page,
+			totalCount,
+		});
+	} catch (err) {
+		res.status(500).send('Server Error');
+	}
+});
+
+router.get('/products/:product', (req, res) => {
+	Product.findById(req.params.product)
+		.then((product) => {
+			if (!product) {
+				return res.status(404).send('Product not found');
+			}
+
+			res.status(200).send(product);
+		})
+		.catch((err) => {
+			if (err.name === 'CastError') {
+				return res.status(404).send('Product not found');
+			}
+			res.status(500).send('Server Error');
+		});
+});
+
+router.get('/products/:product/reviews', async (req, res) => {
+	const productId = req.params.product;
+	const page = parseInt(req.query.page) || 1;
+	const reviewsPerPage = 4;
+	const skip = (page - 1) * reviewsPerPage;
+
+	try {
+		const product = await Product.findById(productId).populate({
+			path: 'reviews',
+			options: {
+				skip: skip,
+				limit: reviewsPerPage,
+			},
+		});
+
+		if (!product) {
+			return res.status(404).send('Product not found');
+		}
+		const totalReviews = await Review.countDocuments({
+			product: productId,
+		});
+
+		res.status(200).json({
+			reviews: product.reviews,
+			currentPage: page,
+			totalPages: Math.ceil(totalReviews / reviewsPerPage),
+		});
+	} catch (err) {
+		if (err.name === 'CastError') {
+			return res.status(404).send('Product not found');
+		}
+		res.status(500).send('Server Error');
+	}
+});
+
+router.post('/products', async (req, res) => {
+	const { category, name, price, image } = req.body;
+
+	if (!category || !name || !price || !image) {
+		return res.status(400).send('All fields are required');
+	}
+
+	try {
+		const newProduct = new Product({
+			category,
+			name,
+			price,
+			image,
+		});
+
+		const savedProduct = await newProduct.save();
+
+		res.status(200).send(savedProduct);
+	} catch (err) {
+		res.status(500).send('Server Error');
```

Comment:
in general we want to do something with the server error, if you are not sending it to the client at least log it into the console.